Model/src/main/java/org/gusdb/wdk/model/user/dataset/event/repo/UserDatasetEventRepo.java

```diff
@@ -0,0 +1,440 @@
+
+    LOG.debug("Fetched {} recovered events", ret.size());
+
+    return ret;
+  }
+
+  public Optional<Long> getEarliestCleanupCompleteEvent() {
+    var sql = "SELECT MIN(event_id) FROM " + schema + TABLE_USER_DATASET_EVENT
+      + " WHERE status = '" + UserDatasetEventStatus.CLEANUP_COMPLETE + "'";
+
+    return new SQLRunner(ds, sql)
+      .executeQuery(rs -> rs.next() ? Optional.of(rs.getLong(1)) : Optional.empty());
+  }
+
+  /**
+   * Retrieves a list of EventRows in the DB that are currently in the
+   * {@code CLEANUP_READY} status.
+   * <p>
+   * This method returns a snapshot at the time of the call and makes no
+   * guarantee that any individual row will be in the {@code CLEANUP_READY}
+   * status once it is time to process that event.  The row should be locked
+   * before processing using the {@link #lockCleanupEvent(EventRow)} method.
+   *
+   * @return A list of zero or more event rows that were in the
+   * {@code CLEANUP_READY} status as of the time this method was called.
+   */
+  public List<EventRow> getCleanupReadyEvents() {
+    var sql = "SELECT\n"
+      + "  event_id\n"
+      + ", user_dataset_id\n"
+      + ", handled_time\n"
+      + ", type_name\n"
+      + ", type_version\n"
+      + "FROM " + schema + TABLE_USER_DATASET_EVENT + "\n"
+      + "WHERE status = '" + UserDatasetEventStatus.CLEANUP_READY + "'";
+
+    var ret = new SQLRunner(ds, sql)
+      .executeQuery((SQLRunner.ResultSetHandler<List<EventRow>>) rs -> {
+        var out = new ArrayList<EventRow>();
+
+        while (rs.next()) {
+          out.add(new EventRow(
+            rs.getLong(1),
+            rs.getLong(2),
+            rs.getObject(3, LocalDateTime.class),
+            UserDatasetEventStatus.CLEANUP_READY,
+            new UserDatasetType(rs.getString(4), rs.getString(5))
+          ));
+        }
+
+        return out;
+      });
+
+    LOG.debug("Fetched {} cleanup ready events.", ret.size());
+
+    return ret;
+  }
+
+  /**
+   * Fetches a set of dataset IDs for datasets that have an event in a status
+   * other than {@code COMPLETE} or {@code CLEANUP_COMPLETE}.
+   * <p>
+   * User datasets with an event in any other status either have a failed or
+   * have been "claimed" by another process.
+   *
+   * @return Set of dataset IDs.
+   */
+  public Set<Long> getIgnoredDatasetIDs() {
+    var sql = "SELECT user_dataset_id FROM " + schema + TABLE_USER_DATASET_EVENT
+      + " WHERE status NOT IN ("
+      + "'" + UserDatasetEventStatus.COMPLETE + "',"
+      + "'" + UserDatasetEventStatus.CLEANUP_COMPLETE + "'"
+      + ")";
+
+    var ret = new SQLRunner(ds, sql).executeQuery((SQLRunner.ResultSetHandler<Set<Long>>) rs -> {
+      var out = new HashSet<Long>();
+      while (rs.next())
+        out.add(rs.getLong(1));
+      return out;
+    });
+
+    LOG.debug("Fetched {} ignorable dataset IDs", ret.size());
+
+    return ret;
+  }
+
+  /**
+   * Attempts to lock the given event row (creating it if it wasn't previously
```

Comment:
if this method only relates to "sync" mode, the comment could say that

Reply:
updated